Beta_Integrations/integration-ThreatQ.yml

```diff
@@ -70,30 +70,72 @@ script:
 
     PROXY = load_proxy()
 
+    def get_errors_string_from_bad_request(bad_request_results):
+        errors_list = bad_request_results.json().get("errors", [])
+        errors_string = ""
+        error_num = 1
+        if errors_list:
+            errors_string = "Errors from server: \n"
+            for error in errors_list:
+                errors_string += "Error #{0}: {1}\n".format(error_num, error)
+                error_num += 1
+        return errors_string
+
     # ThreatQ auth based on OAuth 2.0 credential grand method
     def tq_access():
         data = {'grant_type': 'password','email': EMAIL, 'password': PASSWORD, 'client_id': CLIENT_ID}
         access_token_response = requests.post(API_TOKEN_URL, data=data, verify=False, allow_redirects=False)
+
         tokens = json.loads(access_token_response.text)
+        if int(access_token_response.status_code) >= 400:
+            errors_string = get_errors_string_from_bad_request(access_token_response)
+            error_message = "Authentication failed, unable to retrieve an access token.\n {}".format(errors_string)
+            return_error(error_message)
+
+        new_integration_context = {
+            "access_token": tokens['access_token'],
+            "access_token_creation_time": int(time.time()) -1, # decrementing one second to be on the safe side
+            "access_token_expires_in": tokens['expires_in']
+        }
+        demisto.setIntegrationContext(new_integration_context)
         token = tokens['access_token']
         return token
 
-    ACCESS_TOKEN = tq_access()
+
+    def access_token_not_expired():
+            epoch_time_now = time.time()
+            epoch_time_when_token_granted = demisto.getIntegrationContext().get("access_token_creation_time")
+            token_time_until_expiration = demisto.getIntegrationContext().get("access_token_expires_in")
+            return int(epoch_time_now) - int(epoch_time_when_token_granted) < int(token_time_until_expiration)
```

Comment:
Better to put default values to make sure they are not None

Reply:
This function is only called if access_token is not None.
If access_token is not None, the other fields are not None as well, since all the fields are only defined together